quickjs.c

```diff
@@ -537,6 +538,73 @@ typedef enum JSFunctionKindEnum {
     JS_FUNC_ASYNC_GENERATOR = (JS_FUNC_GENERATOR | JS_FUNC_ASYNC),
 } JSFunctionKindEnum;
 
+typedef struct InlineCacheRingItem {
+    JSShape* shape;
+    uint32_t prop_offset;
+} InlineCacheRingItem;
+
+typedef struct InlineCacheRingSlot {
+    JSAtom atom;
+    InlineCacheRingItem buffer[IC_CACHE_ITEM_CAPACITY];
```

Comment:
This is probably better as two arrays, one with the JSShape and the other with the offset, because right now it has a 4 byte gap per entry on 64 bits systems.

For the same reason, it's better to put the atom after the array because that also creates a 4 byte gap.

My back-of-the-napkin math says that brings down `sizeof(InlineCacheRingSlot)` from 140 to 104 bytes (or 144 vs. 108 because of padding when used in an array.)

Changing the number of entries to 4 reduces it further to 56 bytes and that fits snugly in a cache line or two.

You could use bitfields for the atom and index fields because you only need 2 or 3 bits for the latter, and although that brings the size down to 52 bytes, it's effectively still 56 bytes because of padding.

Reply:
Good point, gets down to 56 bytes indeed: https://godbolt.org/z/YrMddbP88

Applied the changes (except bitfields)